File: dsa/week(5)/stl3.js
class PriorityQueue {
    constructor(comparator = (a, b) => b - a) {
        this.heap = [];
        this.comparator = comparator;
    }

    push(val) {
        this.heap.push(val);
        this.heap.sort(this.comparator);
    }

    pop() {
        return this.heap.shift();
    }

    top() {
        return this.heap[0];
    }

    size() {
        return this.heap.length;
    }

    print() {
        console.log(this.heap);
    }
}

// Max-heap
let pqr = new PriorityQueue();
for (let i = 1; i <= 5; i++) {
    pqr.push(i);
}
pqr.print();
console.log(pqr.size(), pqr.top());
pqr.pop();
pqr.print();

// Min-heap
let pq = new PriorityQueue((a, b) => a - b);
for (let i = 1; i <= 5; i++) {
    pq.push(i);
}
pq.print();

// List operations
let li = [];
li.push(10);  // Push back
li.push(20);  // Push back
li.unshift(30);  // Push front
li.unshift(40);  // Push front
li.unshift(50);  // Push front
li.reverse();
console.log("List after reverse:", li);

li.sort((a, b) => a - b);
console.log("List after sorting:", li);

console.log(li.length);  // Size
console.log(li[0]);  // Front
li.shift();  // Pop front
console.log(li[li.length - 1]);  // Back
li.pop();  // Pop back
console.log("Final list:", li);



// Sorting intervals by a custom comparator
const intervals = [
    { st: 6, et: 4 },
    { st: 3, et: 4 },
    { st: 4, et: 6 },
    { st: 8, et: 13 }
];

// Sort intervals by start time (ascending order)
intervals.sort((i1, i2) => i1.st - i2.st);

intervals.forEach(interval => {
    console.log(`${interval.st} : ${interval.et}`);
});

// Sorting and searching in an array
let arr = [1, 3, 2, 5, 7, 6];

// Sort the array in ascending order
arr.sort((a, b) => a - b);
console.log(arr.join(" "));

// Binary search for a specific value
const binarySearch = (arr, target) => {
    let left = 0, right = arr.length - 1;
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        if (arr[mid] === target) return true;
        else if (arr[mid] < target) left = mid + 1;
        else right = mid - 1;
    }
    return false;
};
console.log(binarySearch(arr, 2));

// Lower bound (first element >= target)
const lowerBound = (arr, target) => {
    let left = 0, right = arr.length;
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        if (arr[mid] < target) left = mid + 1;
        else right = mid;
    }
    return left;
};
console.log(lowerBound(arr, 3));

// Upper bound (first element > target)
const upperBound = (arr, target) => {
    let left = 0, right = arr.length;
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        if (arr[mid] <= target) left = mid + 1;
        else right = mid;
    }
    return left;
};
console.log(upperBound(arr, 3));

File: dsa/week(5)/stl3.py
import heapq
from collections import deque

# Max-heap priority queue
pqr = []
for i in range(1, 6):
    heapq.heappush(pqr, -i)  # Use negative to simulate max-heap

def print_priority_queue(pq):
    print([-x for x in pq])  # Convert back to positive values for printing

print_priority_queue(pqr)
print(len(pqr), -pqr[0])  # Size and top element
heapq.heappop(pqr)  # Remove the top element
print_priority_queue(pqr)

# Min-heap priority queue
pq = []
for i in range(1, 6):
    heapq.heappush(pq, i)

print(pq)  # Min-heap is stored as is
print("Min-Heap elements:", [heapq.heappop(pq) for _ in range(len(pq))])  # Print and empty

# List operations using deque
li = deque()
li.append(10)  # Push back
li.append(20)  # Push back
li.appendleft(30)  # Push front
li.appendleft(40)  # Push front
li.appendleft(50)  # Push front
li.reverse()
print("List after reverse:", list(li))

li = sorted(li)  # Sorting
print("List after sorting:", li)

print(len(li))  # Size
print(li[0])  # Front
li.pop(0)  # Pop front
print(li[-1])  # Back
li.pop()  # Pop back
print("Final list:", li)


# Sorting intervals by a custom comparator
intervals = [
    {"st": 6, "et": 4},
    {"st": 3, "et": 4},
    {"st": 4, "et": 6},
    {"st": 8, "et": 13}
]

# Sort intervals by start time (ascending order)
intervals.sort(key=lambda x: x["st"])
for interval in intervals:
    print(f'{interval["st"]} : {interval["et"]}')

# Sorting and searching in an array
arr = [1, 3, 2, 5, 7, 6]
# Sort the array in ascending order
arr.sort()
print(" ".join(map(str, arr)))

# Binary search for a specific value
from bisect import bisect_left, bisect_right

def binary_search(arr, target):
    index = bisect_left(arr, target)
    return index < len(arr) and arr[index] == target
print(binary_search(arr, 2))

# Lower bound (first element >= target)
def lower_bound(arr, target):
    return bisect_left(arr, target)
print(lower_bound(arr, 3))

# Upper bound (first element > target)
def upper_bound(arr, target):
    return bisect_right(arr, target)
print(upper_bound(arr, 3))
